### ISSUE TITLE
Error Could not move (ossec.log) to (logs/2023/Oct/ossec-08.log)  in Windows Footprint tests

### ISSUE THREAD

**Issue by Deblintrake09** [wazuh/wazuh#19603]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 4.6.0-rc1 | wazuh-agent | Agent | Packages | Windows |

## Description

A new error appeared in https://github.com/wazuh/wazuh/issues/19542. This error is similar to that in https://github.com/wazuh/wazuh/issues/19093 but with a different file and error code.

```
wazuh-agent ERROR: Could not move (ossec.log) to (logs/2023/Oct/ossec-08.log) which returned (32)
wazuh-agent ERROR: Couldn't rename 'ossec.log' to 'logs/2023/Oct/ossec-08.log': No such file or directory
```

**Comment by JorgeMarinoDev** [comment]
It is related to [19216](https://github.com/wazuh/wazuh/pull/19216)
The original problem was related to ERROR_ACCESS_DENIED, now the error is ERROR_SHARING_VIOLATION, meaning the problem might be in the source file, not the target file as in a merge operation.
Actually this occurs in log rotation.
Other Windows users outside Wazuh scope, [related](https://www.bleepingcomputer.com/forums/t/185694/error-message-movefile-failed-code-32/) this to Antivirus agents that hold the source file.
The question here is, why an ossec.log file would be locked for reading by another process?
@cborla

**Comment by vikman90** [comment]
@JorgeMarinoDev Agree, it's related to the previous issue but it's not the cause.

## PoC files

### Open a file

#### With `SHARE_DELETE` flag

<details><summary>open-delete.c</summary>
```c
#include <windows.h>
#include <stdio.h>

int main() {
    const char * TARGET = "D:\\source.txt";

    printf("Creating target file\n");

    const DWORD dwDesiredAccess = GENERIC_WRITE;
    const DWORD dwShareMode = FILE_SHARE_DELETE | FILE_SHARE_READ | FILE_SHARE_WRITE;
    const DWORD dwCreationDisposition = CREATE_ALWAYS;
    const DWORD dwFlagsAndAttributes = FILE_ATTRIBUTE_NORMAL;

    HANDLE hFile = CreateFile(TARGET, dwDesiredAccess, dwShareMode, NULL, dwCreationDisposition, dwFlagsAndAttributes, NULL);

    if (hFile == INVALID_HANDLE_VALUE) {
        printf("CreateFile(): %u\n", (unsigned)GetLastError());
        return 1;
    }

    Sleep(60000);

    CloseHandle(hFile);
    printf("OK\n");

    return 0;
}
```

</details>

#### No `SHARE_DELETE` flag

<details><summary>open.c</summary>
```c
#include <windows.h>
#include <stdio.h>

int main() {
    const char * TARGET = "D:\\source.txt";

    printf("Creating target file\n");

    const DWORD dwDesiredAccess = GENERIC_WRITE;
    const DWORD dwShareMode = FILE_SHARE_READ | FILE_SHARE_WRITE;
    const DWORD dwCreationDisposition = CREATE_ALWAYS;
    const DWORD dwFlagsAndAttributes = FILE_ATTRIBUTE_NORMAL;

    HANDLE hFile = CreateFile(TARGET, dwDesiredAccess, dwShareMode, NULL, dwCreationDisposition, dwFlagsAndAttributes, NULL);

    if (hFile == INVALID_HANDLE_VALUE) {
        printf("CreateFile(): %u\n", (unsigned)GetLastError());
        return 1;
    }

    Sleep(60000);

    CloseHandle(hFile);
    printf("OK\n");

    return 0;
}
```

</details>

### Rename a file

#### Move method

<details><summary>move.c</summary>
```c
#include <windows.h>
#include <stdio.h>

int main() {
    const char * SOURCE = "D:\\source.txt";
    const char * TARGET = "D:\\target.txt";

    BOOL r = MoveFileEx(SOURCE, TARGET, MOVEFILE_REPLACE_EXISTING | MOVEFILE_WRITE_THROUGH);

    if (!r) {
        printf("ReplaceFile(): %u\n", (unsigned)GetLastError());
        return 1;
    }

    printf("OK\n");
    return 0;
}
```

</details>

#### Replace method

<details><summary>replace.c</summary>
```c
#include <windows.h>
#include <stdio.h>

int main() {
    const char * SOURCE = "D:\\source.txt";
    const char * TARGET = "D:\\target.txt";

#ifndef NO_CREATE_FILE
    BOOL file_created = FALSE;
    DWORD dwFileAttributes = GetFileAttributes(TARGET);

    if (dwFileAttributes == INVALID_FILE_ATTRIBUTES) {
        printf("Creating target file\n");

        const DWORD dwDesiredAccess = GENERIC_WRITE;
        const DWORD dwShareMode = FILE_SHARE_DELETE | FILE_SHARE_READ | FILE_SHARE_WRITE;
        const DWORD dwCreationDisposition = CREATE_ALWAYS;
        const DWORD dwFlagsAndAttributes = FILE_ATTRIBUTE_NORMAL;

        HANDLE hFile = CreateFile(TARGET, dwDesiredAccess, dwShareMode, NULL, dwCreationDisposition, dwFlagsAndAttributes, NULL);

        if (hFile == INVALID_HANDLE_VALUE) {
            printf("CreateFile(): %u\n", (unsigned)GetLastError());
            return 1;
        }

        CloseHandle(hFile);
        file_created = TRUE;
    }
#endif

    BOOL r = ReplaceFile(TARGET, SOURCE, NULL, 0, NULL, NULL);

    if (!r) {
        printf("ReplaceFile(): %u\n", (unsigned)GetLastError());

#ifndef NO_CREATE_FILE
        if (file_created) {
            printf("Deleting target file as it was created by this program.\n");
            DeleteFile(TARGET);
        }
#endif

        return 1;
    }

    printf("OK\n");
    return 0;
}
```

</details>

## Results

The `SHARE_DELETE` flag corresponds to the _source file_.

|Open w/o delete|Open w/ delete|Move|Replace|Result|Notes|
|:-:|:-:|:-:|:-:|:-:|---|
| ✔️ | | ✔️ | | 🔴 |`ERROR_ACCESS_DENIED`, behavior before 4.6.0.|
| ✔️ | | | ✔️ | 🔴 |`ERROR_SHARING_VIOLATION` |
| | ✔️ | ✔️ | | 🟢 ||
| | ✔️ | | ✔️ | 🔴 |`ERROR_SHARING_VIOLATION`|

## File rename functions

- `MoveFileEx()` raises `ERROR_ACCESS_DENIED` if the **destination** file is opened, no matter its `SHARE_DELETE` flag.
- `ReplaceFile()` raises `ERROR_SHARING_VIOLATION` if the **source** file is opened, no matter its `SHARE_DELETE` flag.

In other words:
- Replacing `fopen()` with `wfopen()` the _ossec.log_ opening function, and changing `ReplaceFile()` back to `MoveFileEx()`, would fix this problem.
- That would break the shared configuration file unmerging operation again.
